juju_verify/verifiers/ovn_central.py

```diff
@@ -0,0 +1,453 @@
+            result.add_partial_result(
+                Severity.FAIL,
+                f"OVN cluster with {all_units} units can not tolerate simultaneous "
+                f"reboot of {units_to_reboot} units.",
+            )
+        else:
+            result.add_partial_result(
+                Severity.OK,
+                f"OVN cluster with {all_units} units can safely tolerate simultaneous "
+                f"reboot of {units_to_reboot} units.",
+            )
+            if units_to_reboot == fault_tolerance:
+                result.add_partial_result(
+                    Severity.WARN,
+                    "While the rebooted units are down, this cluster won't be able to "
+                    "tolerate any more failures.",
+                )
+
+        return result
+
+    def check_downscale(self) -> Result:
+        """Check that removing selected units won't affect cluster's fault tolerance."""
+        all_units = len(self.all_application_units)
+        units_to_remove = len(self.units)
+        original_tolerance = self.cluster_tolerance(all_units)
+        post_removal_tolerance = self.cluster_tolerance(all_units - units_to_remove)
+        base_msg = f"Removing {units_to_remove} units from cluster of {all_units} "
+
+        if original_tolerance < 1:
+            return Result(
+                Severity.FAIL,
+                f"Cluster of {all_units} units has already 0 fault tolerance.",
+            )
+
+        if post_removal_tolerance < 1:
+            return Result(
+                Severity.FAIL, base_msg + "would bring its fault tolerance to 0."
+            )
+
+        if original_tolerance == post_removal_tolerance:
+            result = Result(Severity.OK, base_msg + "won't impact its fault tolerance.")
+        else:
+            result = Result(
+                Severity.WARN,
+                base_msg
+                + f"will decrease its fault tolerance from {original_tolerance}"
+                f" to {post_removal_tolerance}.",
+            )
+
+        return result
+
+    def preflight_checks(self) -> Result:
+        """Run common checks that verify integrity of the OVN cluster.
+
+        These checks should be prerequisite before any further checks are run for both
+        reboot and shutdown actions
+        """
+        charm_supported = self.check_supported_charm_version()
+        if not charm_supported.success:
+            return charm_supported
+
+        return charm_supported + checks_executor(
+            self.check_single_application,
+            self.check_leader_consistency,
+            self.check_uncommitted_logs,
+            self.check_unknown_servers,
+        )
+
+    def verify_reboot(self) -> Result:
+        """Verify that it's safe to reboot selected ovn-central units."""
+        preflight_results = self.preflight_checks()
+        if not preflight_results.success:
+            return preflight_results
+
+        return preflight_results + self.check_reboot()
```

Comment:
I thin we should use `checks_executor` to run all checks, in this case the functionality of `--stop-on-failure` CLI argument will be not broke, but this executor provide also logs and default result output, What do you think?

Reply:
I understand where you are coming from, but in this case, the checks are really dependent on each other. "preflight_checks" ensure that the cluster is in good and consistent state. If it's not, there's no point of continuing with the main check because it won't have reliable data. Example:

Main checks use application unit count to determine minimum required quorum. For cluster of 3, at least 2 units must remain to maintain quorum. However if there were two previously removed units that failed to leave cluster properly, the OVN cluster would still consider itself as cluster of 5 and minimum servers required for the quorum would be 3.

Reply:
Even more simple example is the check that ovn-central charm supports required actions. It's part of the preflight check and if it fails there's no point of continuing with further checks as they all rely on the "cluster-status" action.

Reply:
[Here](https://github.com/canonical/juju-verify/blob/master/juju_verify/verifiers/neutron_gateway.py#L181-L192) is an example how I think it should be used. I'm not saying that checks should be run at once, but each check or list of checks should be run by `checks_executor`.

Reply:
Oh, I see. I misunderstood. Thanks, I pushed the updated version.